File: plugins/plugin-xai/typescript/services/x.service.ts
import { type IAgentRuntime, logger, parseBooleanFromText, Service } from "@elizaos/core";
import { ClientBase } from "../base";
import { XDiscoveryClient } from "../discovery";
import { validateXConfig } from "../environment";
import { XInteractionClient } from "../interactions";
import { XPostClient } from "../post";
import { XTimelineClient } from "../timeline";
import type { IXClient } from "../types";
import { getSetting } from "../utils/settings";

export class XClientInstance implements IXClient {
  client: ClientBase;
  post?: XPostClient;
  interaction?: XInteractionClient;
  timeline?: XTimelineClient;
  discovery?: XDiscoveryClient;

  constructor(runtime: IAgentRuntime, state: Record<string, unknown>) {
    this.client = new ClientBase(runtime, state);

    const postEnabled = parseBooleanFromText(getSetting(runtime, "X_ENABLE_POST"));
    if (postEnabled) {
      logger.info("X posting ENABLED");
      this.post = new XPostClient(this.client, runtime, state);
    }

    const repliesEnabled = getSetting(runtime, "X_ENABLE_REPLIES") !== "false";
    if (repliesEnabled) {
      logger.info("X replies ENABLED");
      this.interaction = new XInteractionClient(this.client, runtime, state);
    }

    const actionsEnabled = getSetting(runtime, "X_ENABLE_ACTIONS") === "true";
    if (actionsEnabled) {
      logger.info("X timeline actions ENABLED");
      this.timeline = new XTimelineClient(this.client, runtime, state);
    }

    const discoveryEnabled =
      getSetting(runtime, "X_ENABLE_DISCOVERY") === "true" ||
      (actionsEnabled && getSetting(runtime, "X_ENABLE_DISCOVERY") !== "false");
    if (discoveryEnabled) {
      logger.info("X discovery ENABLED");
      this.discovery = new XDiscoveryClient(this.client, runtime, state);
    }
  }
}

export class XService extends Service {
  static serviceType = "x";
  capabilityDescription = "Send and receive posts on X (formerly Twitter)";

  public xClient?: XClientInstance;

  static async start(runtime: IAgentRuntime): Promise<XService> {
    const service = new XService();
    service.runtime = runtime;

    await validateXConfig(runtime);
    logger.log("✅ X configuration validated");

    service.xClient = new XClientInstance(runtime, {});
    await service.xClient.client.init();

    if (service.xClient.post) {
      logger.log("📮 Starting X post client...");
      await service.xClient.post.start();
    }

    if (service.xClient.interaction) {
      logger.log("💬 Starting X interaction client...");
      await service.xClient.interaction.start();
    }

    if (service.xClient.timeline) {
      logger.log("📊 Starting X timeline client...");
      await service.xClient.timeline.start();
    }

    if (service.xClient.discovery) {
      logger.log("🔍 Starting X discovery client...");
      await service.xClient.discovery.start();
    }

    logger.log("✅ X service started");
    return service;
  }

  async stop(): Promise<void> {
    if (this.xClient?.post) await this.xClient.post.stop();
    if (this.xClient?.interaction) await this.xClient.interaction.stop();
    if (this.xClient?.timeline) await this.xClient.timeline.stop();
    if (this.xClient?.discovery) await this.xClient.discovery.stop();
    logger.log("X service stopped");
  }
}

File: plugins/plugin-xai/python/elizaos_plugin_xai/services/x_service.py
"""X (Twitter) main service orchestration."""

import logging
from typing import Protocol

logger = logging.getLogger(__name__)


class RuntimeProtocol(Protocol):
    """Protocol for agent runtime interface."""

    agent_id: str

    def get_setting(self, key: str) -> str | None:
        """Get a runtime setting."""
        ...


class XService:
    """X Client Instance - orchestrates all X (formerly Twitter) functionality.

    Components:
    - client: base operations (auth, timeline caching)
    - post: autonomous posting
    - interaction: mentions and replies
    - timeline: actions (likes, reposts, replies)
    - discovery: content discovery and engagement
    """

    service_type = "x"
    capability_description = "Send and receive posts on X (formerly Twitter)"

    def __init__(self) -> None:
        """Initialize the X service."""
        self._runtime: RuntimeProtocol | None = None
        self._is_running = False
        self._post_enabled = False
        self._replies_enabled = False
        self._actions_enabled = False
        self._discovery_enabled = False

    @classmethod
    async def start(cls, runtime: RuntimeProtocol) -> "XService":
        """Start the X service.

        Args:
            runtime: The agent runtime

        Returns:
            The started service instance
        """
        service = cls()
        service._runtime = runtime

        # Check feature flags
        service._post_enabled = runtime.get_setting("X_ENABLE_POST") == "true"
        service._replies_enabled = runtime.get_setting("X_ENABLE_REPLIES") != "false"
        service._actions_enabled = runtime.get_setting("X_ENABLE_ACTIONS") == "true"
        service._discovery_enabled = runtime.get_setting("X_ENABLE_DISCOVERY") == "true" or (
            service._actions_enabled and runtime.get_setting("X_ENABLE_DISCOVERY") != "false"
        )

        if service._post_enabled:
            logger.info("X posting ENABLED")
        if service._replies_enabled:
            logger.info("X replies ENABLED")
        if service._actions_enabled:
            logger.info("X timeline actions ENABLED")
        if service._discovery_enabled:
            logger.info("X discovery ENABLED")

        service._is_running = True
        logger.info("X configuration validated")
        return service

    async def stop(self) -> None:
        """Stop the X service."""
        self._is_running = False
        logger.info("X service stopped")

    @property
    def is_running(self) -> bool:
        """Check if the service is running."""
        return self._is_running

    @property
    def post_enabled(self) -> bool:
        """Check if posting is enabled."""
        return self._post_enabled

    @property
    def replies_enabled(self) -> bool:
        """Check if replies are enabled."""
        return self._replies_enabled

    @property
    def actions_enabled(self) -> bool:
        """Check if timeline actions are enabled."""
        return self._actions_enabled

    @property
    def discovery_enabled(self) -> bool:
        """Check if discovery is enabled."""
        return self._discovery_enabled
